### PR TITLE
[MSFT] Dynamic Instance op

### PR DESCRIPTION
**Pull request by teqdruid**
Represents an instance (as in instance in the instance hierarchy) referred
to henceforth as a dynamic instance. Specified with a path through the
instance hierarchy (which in the future will be replaced with an AppID).
Lowers to a `hw.globalref` but unlike a global ref, does not require all of
the ops participating in the globalref to contain a back pointer attribute.
Allows users to efficiently add placements to a large number of dynamic
instances which happen to map to a small number of static instances by
bulk-adding the necessary `hw.globalref` attributes.

During the lowering, moves the operations in the body to the top level and
gives them the symbol of the globalref which was created to replace the
dynamic instance.

### CODE REVIEW

File: test/Dialect/MSFT/dynamic_instance.mlir [hunk 1]
```diff
@@ -0,0 +1,71 @@
+// RUN: circt-opt %s -verify-diagnostics | circt-opt -verify-diagnostics
+// RUN: circt-opt %s --msft-lower-instances | FileCheck %s
+
+msft.instance.dynamic [#hw.innerNameRef<@deeper::@branch>, #hw.innerNameRef<@shallow::@leaf>, #hw.innerNameRef<@leaf::@module>] {
```
Comment (teqdruid) [inline test/Dialect/MSFT/dynamic_instance.mlir:4]: @mikeurbach What do you think of this?

File: include/circt/Dialect/MSFT/MSFTOps.td [hunk 2]
```diff
@@ -240,3 +242,19 @@ def PhysLocationOp : MSFTOp<"pd.location"> {
     ($ref^)? custom<PhysLoc>($loc) (`path` `:` $subPath^)? attr-dict
   }];
 }
+
+def DynamicInstanceOp : MSFTOp<"instance.dynamic",
+                               [HasParent<"mlir::ModuleOp">, NoTerminator]> {
+
+  let summary = "A module instance in the instance hierarchy";
+  let arguments = (ins NameRefArrayAttr:$appid);
```
Comment (teqdruid) [inline include/circt/Dialect/MSFT/MSFTOps.td:250]: For now, piggybacking on `NameRefArray` for appid.

File: test/Dialect/MSFT/dynamic_instance.mlir (hunk 1 above)
Comment (mikeurbach) [reply test/Dialect/MSFT/dynamic_instance.mlir:4]: I think it looks great. The whole series of patches looks like a really good direction.

File: include/circt/Dialect/MSFT/MSFTOps.td (hunk 2 above)
Comment (mikeurbach) [reply include/circt/Dialect/MSFT/MSFTOps.td:250]: Makes sense. If we can, using the HW attributes should help make it easier to integrate some day.

**Review by mikeurbach — APPROVED** [review]
Looks good, but CI failed for some reason. I think that needs to be addressed, but hopefully is a minor update.

**Comment by teqdruid** [comment]
Yeah, it runs fine on my machine but valgrind shows a bunch of errors.

**Comment by teqdruid** [comment]
Simple fix. Thanks @mikeurbach !

**Pull request merged**